nf_core/modules.py

```diff
@@ -231,3 +237,481 @@ def has_valid_pipeline(self):
+            self.passed.append("meta.yml exists {}".format(file))
+        except FileNotFoundError:
+            self.failed.append("meta.yml doesn't exist for {} ({})".format(module_name, file))
+            return
+
+        # Confirm that all required keys are given
+        contains_required_keys = True
+        for rk in required_keys:
+            if not rk in meta_yaml.keys():
+                self.failed.append("{} not specified in {}".format(rk, file))
+                contains_required_keys = False
+            if contains_required_keys:
+                self.passed.append("{} contains all required keys".format(file))
+
+        # Confirm that all input and output channels are specified
+        meta_input = [list(x.keys())[0] for x in meta_yaml["input"]]
+        for input in inputs:
+            if input in meta_input:
+                self.passed.append("{} specified for {}".format(input, module_name))
+            else:
+                self.failed.append("{} missing in meta.yml for {}".format(input, module_name))
+
+        meta_output = [list(x.keys())[0] for x in meta_yaml["output"]]
+        for output in outputs:
+            if output in meta_output:
+                self.passed.append("{} specified for {}".format(output, module_name))
+            else:
+                self.failed.append("{} missing in meta.yml for {}".format(output, module_name))
+
+        # TODO --> decide whether we want/need this test? or make it silent for now
+        # Check that 'name' adheres to guidelines
+        software_name = file.split("software")[1].split(os.sep)[1]
+        if module_name == software_name:
+            required_name = module_name
+        else:
+            required_name = software_name + " " + module_name
+
+        if meta_yaml["name"] == required_name:
+            self.passed.append("meta.yaml module name is correct: {}".format(module_name))
+        else:
+            self.warned.append("meta.yaml module name not according to guidelines: {}".format(module_name))
+
+    def lint_main_nf(self, file):
+        """
+        Lint a single main.nf module file
+        Can also be used to lint local module files,
+        in which case failures should be interpreted
+        as warnings
+        """
+        inputs = []
+        outputs = []
+
+        # Check whether file exists and load it
+        try:
+            with open(file, "r") as fh:
+                lines = fh.readlines()
+            self.passed.append("Module file exists {}".format(file))
+        except FileNotFoundError as e:
+            self.failed.append("Module file doesn't exist {}".format(file))
+            return
+
+        # Check that options are defined
+        options_keywords = ["def", "options", "=", "initOptions(params.options)"]
+        options_keywords_2 = ["params.options", "=", "[:]"]
+
+        if any(l.split() == options_keywords for l in lines) and any(l.split() == options_keywords_2 for l in lines):
```

Comment:
Is this all just to avoid having to use a regex? 😆 

```suggestion
        initoptions_re = re.compile(r"\s*def\s+options\s*=\s*initOptions\s*\(\s*params\.options\s*\)\s*")
        paramsoptions_re = re.compile(r"\s*params\.options\s*=\s*\[:\]\s*")
        if any(initoptions_re.match(l) for l in lines) and any(paramsoptions_re.match(l) for l in lines):
```

Although this is a little less pretty I think it will be a lot more tolerant of whitespace.

Reply:
You got me :D well no actually I was going through the code looking for where regexes could be helpful, and I just didn't see it here 🤷 
I guess leading whitespaces could potentially lead to problems, as well as anything written behind the the lines (e.g. comments) - so yep thinking about it the regex thing is more robust - thanks!